modules/heat_conduction/src/materials/AnisoHeatConductionMaterial.C

```diff
@@ -10,137 +10,152 @@
-    _thermal_conductivity_y(isParamValid("thermal_conductivity_y") ||
-                                    isParamValid("thermal_conductivity_y_pp")
-                                ? &declareProperty<Real>("thermal_conductivity_y")
-                                : NULL),
-    _thermal_conductivity_y_dT(
-        _thermal_conductivity_y ? &declareProperty<Real>("thermal_conductivity_y_dT") : NULL),
-    _thermal_conductivity_z(isParamValid("thermal_conductivity_z") ||
-                                    isParamValid("thermal_conductivity_z_pp")
-                                ? &declareProperty<Real>("thermal_conductivity_z")
-                                : NULL),
-    _thermal_conductivity_z_dT(
-        _thermal_conductivity_z ? &declareProperty<Real>("thermal_conductivity_z_dT") : NULL),
-
-    _specific_heat(declareProperty<Real>("specific_heat")),
+template <bool is_ad>
+AnisoHeatConductionMaterialTempl<is_ad>::AnisoHeatConductionMaterialTempl(
+    const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _ref_temp(getParam<Real>("reference_temperature")),
+
+    _has_temp(isCoupled("temperature")),
+    _T(coupledGenericValue<is_ad>("temperature")),
+    _T_var(coupled("temperature")),
+    _T_name(getVar("temperature", 0)->name()),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+
+    _user_provided_thermal_conductivity(getParam<std::vector<Real>>("thermal_conductivity")),
+    _user_provided_specific_heat(isParamValid("specific_heat") ? getParam<Real>("specific_heat")
+                                                               : 500),
+
+    _thermal_conductivity(
+        declareGenericProperty<RankTwoTensor, is_ad>(_base_name + "thermal_conductivity")),
+    _dthermal_conductivity_dT(
+        declarePropertyDerivative<RankTwoTensor>(_base_name + "thermal_conductivity", _T_name)),
+    _thermal_conductivity_temperature_coefficient_function(
+        getParam<FunctionName>("thermal_conductivity_temperature_coefficient_function") != ""
+            ? &getFunction("thermal_conductivity_temperature_coefficient_function")
+            : nullptr),
+    _specific_heat(declareGenericProperty<Real, is_ad>(_base_name + "specific_heat")),
+    _dspecific_heat_dT(declarePropertyDerivative<Real>(_base_name + "specific_heat", _T_name)),
     _specific_heat_temperature_function(
         getParam<FunctionName>("specific_heat_temperature_function") != ""
             ? &getFunction("specific_heat_temperature_function")
-            : NULL)
+            : nullptr)
 {
-  bool k_x = isParamValid("thermal_conductivity_x") || (NULL != _thermal_conductivity_x_pp);
-  bool k_y = isParamValid("thermal_conductivity_y") || (NULL != _thermal_conductivity_y_pp);
-  bool k_z = isParamValid("thermal_conductivity_z") || (NULL != _thermal_conductivity_z_pp);
+  if (!isParamValid("specific_heat") && !_specific_heat_temperature_function)
```

Comment:
this is an antipattern. You can make `specific_heat` a function argument and just drop `specific_heat_temperature_function`. Users can always supply a number literal where a function is expected. That would get rid of some of the messy logic.

Reply:
Hmmm, looking below I can see you might want to avoid calling the setDerivative function for constant specific heat. That might be a premature optimization though.

Reply:
The user might want to define specific heat as a function of temperature (https://github.com/idaholab/moose/discussions/18082), so I would leave it there for completeness. Let me know if you think otherwise and I'll change it.

Reply:
Yeah, I know. What I'm saying is that both cases could be captured by _just_ using a Function parameter. It is perfectly legal to supply a number as the value for a function parameter. MOOSE will generate an anonymous ConstantFunction object behind the scenes that will always evaluate to the specified value.

But as I noted below there is a considerable overhead since you're setting derivatives if the user specifies a non-constant value.

The optimimum solution IMO would be to hide this complexity from the user by checking if the returned function is of type `ConstantFunction` using a `dynamic_cast<const ConstantFunction *>` and assign the value for the specific heat on a single evaluation of that function and then do what you did below (i.e. not setting the derivatives).

Reply:
Done.